Ghidra/Processors/Xtensa/data/languages/xtensaInstructions.sinc

```diff
@@ -0,0 +1,1283 @@
+}
+
+# SSA8B - Set Shift Amount for BE Byte Shift, pg. 531.
+:ssa8b as is op2 = 0b0100 & op1 = 0 & ar = 0b0011 & as & at = 0 & op0 = 0 {
+    local lsa:4 = (as&3)*8;
+    sar = 32 - lsa:1;
+}
+
+# SSA8L - Set Shift Amount for LE Byte Shift, pg. 532.
+:ssa8l as is op2 = 0b0100 & op1 = 0 & ar = 0b0010 & as & at = 0 & op0 = 0 {
+    local rsa:4 = (as&3)*8;
+    sar = rsa:1;
+}
+
+# SSAI - Set Shift Amount Immediate, pg. 533.
+:ssai u5_8.11_4 is op2 = 0b0100 & op1 = 0 & ar = 0b0100 & u5_8.11_4 & u3_5.7 = 0 & op0 = 0 {
+    sar = u5_8.11_4:1;
+}
+
+# SSI - Store Single Immediate, pg. 534.
+:ssi ft, as, u10_16.23_sb2 is u10_16.23_sb2 & ar = 0b0100 & as & ft & op0 = 0b0011 {
+    local addr:4 = as + zext(u10_16.23_sb2);
+    *:4 addr = ft;
+}
+
+# SSIU - Store Single Immediate with Update, pg. 536.
+:ssiu ft, as, u10_16.23_sb2 is u10_16.23_sb2 & ar = 0b1100 & as & ft & op0 = 0b0011 {
+    local addr:4 = as + zext(u10_16.23_sb2);
+    *:4 addr = ft;
+    as = addr;
+}
+
+# SSL - Set Shift Amount for Left Shift, pg. 538.
+:ssl as is op2 = 0b0100 & op1 = 0 & ar = 0b0001 & as & at = 0 & op0 = 0 {
+    sar = 32 - (as:1 & 0xf);
+}
+
+# SSR - Set Shift Amount for Right Shift, pg. 539.
+:ssr as is op2 = 0b0100 & op1 = 0 & ar = 0 & as & at = 0 & op0 = 0 {
+    sar = (as:1 & 0xf);
+}
+
+# SSX - Store Singe Indexed, pg. 540.
+:ssx fr, as, at is op2 = 0b0100 & op1 = 0b1000 & fr & as & at & op0 = 0 {
+    local addr:4 = as+at;
+    *:4 addr = fr;
+}
+
+# SSXU - Store Singe Indexed with Update, pg. 541.
+:ssxu fr, as, at is op2 = 0b0101 & op1 = 0b1000 & fr & as & at & op0 = 0 {
+    local addr:4 = as+at;
+    *:4 addr = fr;
+    as = addr;
+}
+
+# SUB - Subtract, pg. 542.
+:sub ar, as, at is op2 = 0b1100 & op1 = 0 & ar & as & at & op0 = 0 {
+    ar = as - at;
+}
+
+# SUB.S - Subtract Single, pg. 543.
+:sub.s fr, fs, ft is op2 = 0b0001 & op1 = 0b1010 & fr & fs & ft & op0 = 0 {
+    fr = fs f- ft;
+}
+
+# SUBX2 - Subtract with Shift by 1, pg. 544.
+:subx2 ar, as, at is op2 = 0b1101 & op1 = 0 & ar & as & at & op0 = 0 {
+    ar = (as << 1) - at;
+}
+
+# SUBX4 - Subtract with Shift by 2, pg. 545.
+:subx4 ar, as, at is op2 = 0b1110 & op1 = 0 & ar & as & at & op0 = 0 {
+    ar = (as << 2) - at;
+}
+
+# SUBX8 - Subtract with Shift by 3, pg. 546.
+:subx8 ar, as, at is op2 = 0b1111 & op1 = 0 & ar & as & at & op0 = 0 {
+    ar = (as << 3) - at;
+}
+
+# SYSCALL - System Call, pg. 547.
+:syscall is op2 = 0 & op1 = 0 & ar = 0b0101 & as = 0 & at = 0 & op0 = 0 {
+    syscall();
+}
+
+# TRUNC.S - Truncate Single to Fixed, pg. 548
+:trunc.s ar, fs, u4_4.7 is op2 = 0b1001 & op1 = 0b1010 & ar & fs & u4_4.7 & op0 = 0 {
+    local scale:4 = int2float(1:2 << u4_4.7:2);
+    ar = trunc(fs f* scale);
+}
+
+# UEQ.S - Compare Single Unordered or Equal, pg. 549.
+:ueq.s br, fs, ft is op2 = 0b0011 & op1 = 0b1011 & br & fs & ft & op0 = 0 {
+    br = nan(fs) || nan(ft) || fs f== ft;
+}
+
+# UFLOAT.S - Convert Unsigned Fixed to Single, pg. 550. XXX: How is this different from float.as?
```

Comment:
I had the same question when i came across this type of function, and not sure how the PCODE op handles signed vs unsigned here, i think i left as a todo or a note, but along the lines of:
```
local tmp:8 = zext(as);
local tmpf:8 = int2float(tmp);
local f:4 = float2float(tmpf);
...
```

Reply:
Afaik int2float and float2float just makes the decompiler typecast the variable. I once did a test with a bunch of fixed floating point conversion instructions where I just had them as float2float and I ended up with something like:
```
float fVar;

fVar = (float)(float)(float)(float)(float)(float)(float)fVar;
```

So with the small example you have it'll likely produce something like:
```
uchar uVar1;
double dVar1;
float fVar1;

uVar1 = as;
dVar1 = (double)(float)(int) uVar1;
fVar1 = (float) dVar1;
```
Or something along those lines depending on how much the decompiler knows. I'm not 100% certain about the double cast though.

Reply:
@astrelsky  I think you missed the entire point of my comment addressing the comment on that line: how to convert 0xFFFFFFFF into a float that presents MAX_UINT and not -1.

Reply:
> @astrelsky I think you missed the entire point of my comment addressing the comment on that line: how to convert 0xFFFFFFFF into a float that presents MAX_UINT and not -1.
> 
> this should work, i just had in my head they need to be the same sized.
> 
> ```
> local tmp:8 = zext(as);
> local f:4 = int2float(tmp);
> ```

Indeed I did miss it.

I tracked down the decompiler code for the int4float opcode.
https://github.com/NationalSecurityAgency/ghidra/blob/02df944b0e62c6d94f55f93b6682d70a444c705c/Ghidra/Features/Decompiler/src/decompile/cpp/float.cc#L490-L500

Reply:
Done, thanks.

Do you have any suggestion on how to handle :utrunc.s? I could do a `local tmp:8 = int2float(fs f* scale)` as well, but then I’d have to handle overflow and underflow myself, I think?

Reply:
ah, yeah... that looks like it could be annoying:
```
local ftmp = int2float(fs f* scale);
local posov = ?;
local negov = ?;
local ncond = (inf(ftmp) & ftmp f< 0) | negov; 
local pcond = (inf(ftmp) & ftmp f> 0) | nan(ftmp) | posov;
local cond = !ncond & !pcond;
// pos ov, pos inf, NaN -> 0x7fffffff
// neg ov, neg inf -> 0x80000000  
as  = (zext(pcond) * 0x7fffffff) + (zext(ncond) * 0x80000000) + (zext(cond) * trunc(ftmp));
```
you'd have to check the IEEE spec for inf (but maybe #786 someday) and your logic for overflow... and then choose between the multiply cond/add approach or using gotos.  You also have the option of leaving as is since that might be a corner case that never hits (with a note/todo maybe) and it will get implemented when ov/inf/nan check is needed